src/java/htsjdk/samtools/DownsamplingIterator.java

```diff
@@ -123,6 +152,45 @@ private boolean advance() {
         return this.nextRecord != null;
     }
 
+    /**
+     * Buffers reads until either the end of the file is reached or enough reads have been buffered that
+     * downsampling can be performed and the target accuracy achieved.  Once reads have been buffered,
+     * template names are randomly sampled out for discarding until the desired number of reads have
+     * been discarded.
+     *
+     * @return True if one or more reads have been buffered, false otherwise
+     */
+    private boolean bufferNextChunkOfRecords() {
+        final int templatesToRead = (int) Math.ceil(1 / this.targetAccuracy);
+        final Set<String> names = new HashSet<String>();
+        final List<SAMRecord> recs = new ArrayList<SAMRecord>(templatesToRead);
+
+        while (this.underlyingIterator.hasNext() && names.size() < templatesToRead) {
+            final SAMRecord rec = this.underlyingIterator.next();
+            recs.add(rec);
+
+            if (!this.allowSecondaryAlignments && rec.getNotPrimaryAlignmentFlag()) continue;
+            if (!this.allowSupplementalAlignments && rec.getSupplementaryAlignmentFlag()) continue;
+            if (this.decisions.containsKey(rec.getReadName())) continue;
+            names.add(rec.getReadName());
+        }
+
+        final int templatesRead   = names.size();
+        final double rawTemplatesToKeep = templatesRead * this.probabilityOfKeeping;
+        final int templatesToKeep = (keptTemplates / (double) totalTemplates < this.probabilityOfKeeping)
+                ? (int) Math.ceil(rawTemplatesToKeep) : (int) Math.floor(rawTemplatesToKeep);
+        final List<String> tmp    = new ArrayList<String>(names);
+        while (names.size() > templatesToKeep) {
+            names.remove(tmp.get(this.random.nextInt(templatesRead)));
```

Comment:
This could take a while if you are downsampling a large amount (as you'll be hitting already-removed names). Could you make tmp a linkedlist and also remove the drawn integer position from the list? Otherwise, you could hash the names using murmur or similar and take the first templatesToKeep by hash order.